File: algorithms/dynamic_programming/rotateArrayForMaxValue.js
// leetcode 396

const rotateArrayMaxForMaxValue = (A) => {
  let len_A = A.length;
  let sum_A = 0;
  let product_sum = 0;

  for (let i=0; i < len_A; i++) {
    sum_A += A[i];
    product_sum += A[i] * i;
  }
  let max = product_sum;
  for (let j=len_A-1; j > 0; j--) {
    product_sum += sum_A - (A[j] * len_A);
    max = Math.max(product_sum, max);
  }
  return max
}

console.log(rotateArrayForMaxValue([3,4,2,1,2]));
    
    

File: algorithms/dynamic_programming/rotateArrayForMaxValue.py
# leetcode 396 (rotate-function)

def rotateArrayForMaxValue(A):
    len_A = len(A)
    sum_A = products_sum = 0

    for index, value in enumerate(A):
        sum_A += value
        products_sum += index * value
    max_sum = products_sum

    for biggest in range(len_A-1, 0, -1):
        products_sum += sum_A - (len_A * A[biggest])
        max_sum = max(products_sum, max_sum)
        print(products_sum)



